data/account/participationRegistry.go

```diff
@@ -0,0 +1,385 @@
+		}
+
+		// Delete rows
+
+		_, err = tx.Exec(deleteKeysets, pk)
+		if err != nil {
+			return err
+		}
+
+		_, err = tx.Exec(deleteRolling, pk)
+		if err != nil {
+			return err
+		}
+
+		return nil
+	})
+}
+
+func scanRecords(rows *sql.Rows) ([]ParticipationRecord, error) {
+	results := make([]ParticipationRecord, 0)
+	for rows.Next() {
+		var record ParticipationRecord
+		var participationBlob []byte
+		var accountBlob []byte
+		err := rows.Scan(
+			&participationBlob,
+			&accountBlob,
+			&record.FirstValid,
+			&record.LastValid,
+			&record.KeyDilution,
+			&record.LastVote,
+			&record.LastBlockProposal,
+			&record.LastCompactCertificate,
+			&record.EffectiveFirst,
+			&record.EffectiveLast,
+		)
+		if err != nil {
+			return nil, err
+		}
+
+		copy(record.ParticipationID[:], participationBlob)
+		copy(record.Account[:], accountBlob)
+
+		results = append(results, record)
+	}
+
+	return results, nil
+}
+
+func (db *participationDB) Get(id ParticipationID) (record ParticipationRecord, err error) {
+	err = db.store.Rdb.Atomic(func(ctx context.Context, tx *sql.Tx) error {
+		rows, err := tx.Query(selectRecord, id[:])
+		if err != nil {
+			return fmt.Errorf("unable to scan record: %w", err)
+		}
+
+		records, err := scanRecords(rows)
+		if err != nil {
+			return fmt.Errorf("unable to scan record: %w", err)
+		}
+
+		if len(records) != 1 {
+			return fmt.Errorf("expected 1 result found %d", len(records))
+		}
+
+		record = records[0]
+
+		return nil
+	})
+
+	return
+}
+
+func (db *participationDB) GetAll() (records []ParticipationRecord, err error) {
+	err = db.store.Rdb.Atomic(func(ctx context.Context, tx *sql.Tx) error {
+		rows, err := tx.Query(selectRecords)
+		if err != nil {
+			return fmt.Errorf("unable to query records: %w", err)
+		}
+
+		records, err = scanRecords(rows)
+		if err != nil {
+			records = nil
+			return fmt.Errorf("problem scanning records: %w", err)
+		}
+
+		return nil
+	})
+
+	return
+}
+
+func (db *participationDB) Register(id ParticipationID, on basics.Round) error {
+	// Lookup record for first/last valid and account.
+	record, err := db.Get(id)
+	if err != nil {
+		return fmt.Errorf("unable to lookup id: %w", err)
+	}
+
+	// round out of valid range.
+	if on+maxBalLookback > record.LastValid || on+maxBalLookback < record.FirstValid {
+		return ErrInvalidRegisterRange
+	}
+
+	return db.store.Wdb.Atomic(func(ctx context.Context, tx *sql.Tx) error {
+		// if the is an active key, shut it down.
+		_, err = tx.Exec(clearRegistered, on+maxBalLookback, record.Account[:])
+		if err != nil {
+			return fmt.Errorf("unable to clear registered key: %w", err)
+		}
+
+		// update id
+		_, err = tx.Exec(setRegistered, on+maxBalLookback, record.LastValid, id[:])
+		if err != nil {
+			return fmt.Errorf("unable to update registered key: %w", err)
+		}
+
+		return nil
+	})
+}
+
+func (db *participationDB) Record(account basics.Address, round basics.Round, participationAction ParticipationAction) error {
```

Comment:
don't forget that you'd want to make this into a non-blocking : missing a database update because the disk access is slow, is fine. But you don't want to slow down the agreement due to slow disk performance.

Reply:
Do you think running this in a goroutine is sufficient? It looks like the synchronization options have a non-zero chance of DB corruption.

Reply:
@tsachiherman I decided to keep it synchronous at this point and name the agreement abstraction `RecordAsync`. The node layer makes it non-blocking.

Reply:
I think that you'd rather do it here so that you can update the in-memory statistics without blocking, while allowing the persistence take longer.
i.e. Ideally, you would be able to respond to a http query without any blocking on the database.